File: pg-index-health-spring-boot-demo/src/test/java/io/github/mfvanek/pg/index/health/demo/controller/DbStatisticsControllerMockTest.java
/*
 * Copyright (c) 2019-2024. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/pg-index-health-spring-boot-demo
 *
 * Licensed under the Apache License 2.0
 */

package io.github.mfvanek.pg.index.health.demo.controller;

import io.github.mfvanek.pg.common.management.DatabaseManagement;
import io.github.mfvanek.pg.index.health.demo.utils.BasePgIndexHealthDemoSpringBootTest;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.Mockito;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;

import static org.assertj.core.api.Assertions.assertThat;

class DbStatisticsControllerMockTest extends BasePgIndexHealthDemoSpringBootTest {

    @MockBean
    private DatabaseManagement databaseManagement;

    @ParameterizedTest
    @ValueSource(booleans = {false, true})
    void shouldReturnErrorWhenResetStatisticsUnsuccessful(final boolean wait) {
        Mockito.when(databaseManagement.resetStatistics())
            .thenReturn(Boolean.FALSE);
        final Object result = webTestClient.post()
            .uri(uriBuilder -> uriBuilder
                .pathSegment("db", "statistics", "reset")
                .build())
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(wait)
            .exchange()
            .expectStatus().isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR)
            .expectBody(Object.class)
            .returnResult()
            .getResponseBody();
        assertThat(result)
            .isNotNull()
            .satisfies(b -> assertThat(b.toString()).contains("Internal Server Error"));
    }
}

File: pg-index-health-spring-boot-kotlin-demo/src/test/kotlin/io/github/mfvanek/pg/index/health/demo/kotlin/controller/DbStatisticsControllerMockTest.kt
/*
 * Copyright (c) 2019-2025. Ivan Vakhrushev and others.
 * https://github.com/mfvanek/pg-index-health-demo
 *
 * Licensed under the Apache License 2.0
 */

package io.github.mfvanek.pg.index.health.demo.kotlin.controller

import io.github.mfvanek.pg.health.checks.management.DatabaseManagement
import io.github.mfvanek.pg.index.health.demo.kotlin.utils.BasePgIndexHealthDemoSpringBootTest
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mockito.`when`
import org.springframework.boot.test.system.OutputCaptureExtension
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean

@ExtendWith(OutputCaptureExtension::class)
class DbStatisticsControllerMockTest : BasePgIndexHealthDemoSpringBootTest() {

    @MockitoBean
    private lateinit var databaseManagement: DatabaseManagement

    @BeforeEach
    fun setUp() {
        `when`(databaseManagement.resetStatistics()).thenReturn(true)
    }

    @Test
    fun shouldThrowExceptionWhenResetStatisticsWithoutWaitFails() {
        `when`(databaseManagement.resetStatistics()).thenReturn(false)

        webTestClient.post()
            .uri("/db/statistics/reset")
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(false)
            .headers(this::setUpBasicAuth)
            .exchange()
            .expectStatus().is5xxServerError
            .expectBody()
            .jsonPath("$.message").isEqualTo("Statistics reset failed: Could not reset statistics")
    }
}
